File: Python/Algorithms/566.py
from itertools import chain


class Solution:
    def matrixReshape(self, nums, r, c):
        rows, cols = len(nums), len(nums[0])

        # can't fit in new row/col orientation.
        if (r * c) > (rows * cols):
            return nums

        # go through elements of nums by chaining
        it, next_ = chain.from_iterable(nums), next
        result = []
        for i in range(r):
            sub = []
            for j in range(c):
                sub.append(next_(it))
            result.append(sub)
        return result

File: Javascript/Algorithms/566.js
let matrixReshape = (nums, r, c) => {
    const rows = nums.length,
          cols = nums[0].length;
    
    if ((r * c) > (rows * cols)){
        return nums;
    }

    const result = [];
    let k = 0, z = 0;  
    // Fill from nums.
    for(let i = 0; i < r; i++){
        let sub = [];
        for (let j = 0; j < c; j++){
            sub.push(nums[k][z++]);
            if (z == cols){
                z = 0;
                k++;
            }
        }
        result.push(sub);
    }
    return result;
};
